firedrake/extrusion_numbering.pyx

```diff
@@ -0,0 +1,1035 @@
+                                               layer_extents[:, 0])
+
+    boundary_nodes = numpy.empty(maxsize, dtype=IntType)
+    idx = 0
+    for f in range(nfacet):
+        if all_facets:
+            facet = f
+        else:
+            facet = facet_indices[f]
+        local_facet = local_facets[facet]
+        layers = layer_extents[f, 1] - layer_extents[f, 0]
+
+        for i in range(nlocal):
+            dof = local_nodes[local_facet, i]
+            for j in range(layers - 1):
+                boundary_nodes[idx] = facet_node_list[facet, dof] + j * offsets[dof]
+                idx += 1
+
+    return numpy.unique(boundary_nodes[:idx])
+
+
+@cython.wraparound(False)
+def cell_entity_masks(mesh):
+    """Compute a masking integer for each cell in the extruded mesh.
+
+    This integer indicates for each cell, which topological entities
+    in the cell are on the boundary of the domain.  If the ith bit in
+    the integer is on, that indicates that the ith entity is on the
+    boundary, meaning that the appropriate boundary mask should be
+    used to discard element tensor contributions when assembling
+    bilinear forms.
+
+    :arg mesh: the extruded mesh.
+    :returns: a tuple of section, bottom, and top masks.  The section
+        records the number of entities in each column and the offset
+        in the masking arrays for the start of each column.
+    """
+    cdef:
+        numpy.ndarray[PetscInt, ndim=2, mode="c"] cell_closure = mesh.cell_closure
+        numpy.ndarray[PetscInt, ndim=2, mode="c"] layer_extents = mesh.layer_extents
+        numpy.ndarray[numpy.int64_t, ndim=1, mode="c"] top_mask, bottom_mask
+        numpy.ndarray[numpy.int64_t, ndim=1, mode="c"] flips
+        numpy.ndarray[PetscInt, ndim=1, mode="c"] points
+        int cell, ncell, nclosure, idx
+        PetscInt cStart, cEnd
+        PETSc.Section section
+        PetscInt c, cell_bottom, cell_top, layer, p, e, ent, point
+        PetscInt ent_top, ent_bottom
+
+    ncell = cell_closure.shape[0]
+    nclosure = cell_closure.shape[1]
+    cStart, cEnd = mesh._plex.getHeightStratum(0)
+    top_mask = numpy.zeros(numpy.sum(layer_extents[cStart:cEnd, 1] - layer_extents[cStart:cEnd, 0] - 1),
+                           dtype=numpy.int64)
+    bottom_mask = numpy.zeros_like(top_mask)
+    # We iterate over the base cell and do all the entities in the extruded cell above it,
+    # therefore we need a mapping from the standard firedrake entity ordering to this one.
+    fiat_cell = as_fiat_cell(mesh.ufl_cell())
+    points = eutils.entity_reordering(fiat_cell)
+    # Some masks incorporate others (since the closure of the entity
+    # is not just the entity itself), we'll use this to minimise the
+    # number of mask bits that are on.
+    flips = numpy.zeros(points.shape[0], dtype=numpy.int64)
+    for k, v in eutils.entity_closures(fiat_cell).items():
+        flips[k] = ~sum(2**x for x in v)
```

Comment:
So the idea here is that we need to maintain, on each cell of the extruded mesh, information that indicates which topological entities are "exposed" in that cell (either to the top or the bottom of the mesh).  That's one bit of information per topological entity.  Fortunately, all the cells we deal with have at most 27 topological entities (the hexahedron).  So I could use a 32bit int to encode this.  But wait, interior facets need twice as much, so I use a 64bit int.

Moreover, to minimise the amount of duplicate masking we do in the generated code, we notice that we only need to record the "largest" entity that is exposed (since the closure of that will pick up all the others).  So `flips[k]` stores the bitmask that turns off all entities in the closure of k.